upf/substituter.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2021 AIPlan4EU project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from upf.exceptions import UPFTypeError
+from upf.expression import Expression
+import upf.environment
+from upf.walkers.identitydag import IdentityDagWalker
+import upf.walkers as walkers
+import upf.operators as op
+from upf.fnode import FNode
+from typing import List
+import typing
+from upf.type_checker import TypeChecker
+
+class Substituter(IdentityDagWalker):
+    """Performs substitution into an expression """
+    def __init__(self, env: 'upf.environment.Environment'):
+        IdentityDagWalker.__init__(self, env, True)
+        self.env = env
+        self.manager = env.expression_manager
+        self.type_checker = TypeChecker(env)
+
+    def _get_key(self, expression, **kwargs):
+        return expression
+
+    def substitute(self, expression: FNode, substitutions: typing.Dict[Expression, Expression] = {}) -> FNode:
+        """Performs substitution into the given expression.
+
+        Lets consider the examples:
+        f = a & b
+        subs = {a -> c, (c & b) -> d, (a & b) -> c}
+        substitute(f, subs) = c
+
+        f = a
+        subs = {a -> c, c -> d}
+        substitute(f, subs) = c
+
+        f = a & b
+        subs = {a -> 5, b -> c}
+        substitute(f, subs) raises an UPFTypeError
+
+        Note that, since subs is a dictionary:
+        f = a
+        subs = {a -> b, a -> c}
+        substitute(f, subs) = c
+        """
+
+        if len(substitutions) == 0:
+            return expression
+        new_substitutions: typing.Dict[FNode, FNode] = {}
+        new_keys = self.manager.auto_promote(substitutions.keys())
+        new_values = self.manager.auto_promote(substitutions.values())
+        new_substitutions = dict(zip(new_keys, new_values))
+        for k, v in new_substitutions.items():
+            if not self.type_checker.is_compatible_type(k, v):
+                raise UPFTypeError(
+                    f"The expression type of {str(k)} is not compatible with the given substitution {str(v)}")
+        return self.walk(expression, subs = new_substitutions)
```

Comment:
This part is a bit too convoluted: why not simply iterate over `substitutions` and create `new_substitution and do the type check on-the-fly?

Reply:
@mikand  This idea is because auto_promote can take a list and gives always back a list, so calling it on one single argument would mean to create a list for just one argument.

I don't really know if this has a highly negative impact or it's something negligible.

Reply:
One clever solution I can find is:
`
        new_keys = self.manager.auto_promote(substitutions.keys())
        new_values = self.manager.auto_promote(substitutions.values())
        for k, v in zip(new_keys, new_values):
               if self.type_checker.is_compatible_type(k, v):
                        new_substitutions[k] = v
                else:
                        raise UPFTypeError(...)
`

Reply:
I actually missed the part "iterate over `substitution`", my brain red "iterate over `new_substitution`".

Ok!